File: src/main/java/com/adinsight/agent/session/AgentSession.java
package com.adinsight.agent.session;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.OptionalLong;
import java.util.UUID;

public class AgentSession {
    private final String sessionId;
    private final String userId;
    private final List<AgentTurn> turns;
    private final Map<String, Object> state;
    private int persistedTurnCount;

    public AgentSession(String sessionId, String userId) {
        this(sessionId, userId, new ArrayList<>(), new HashMap<>(), 0);
    }

    public AgentSession(
            String sessionId,
            String userId,
            List<AgentTurn> turns,
            Map<String, Object> state,
            int persistedTurnCount
    ) {
        this.sessionId = sessionId;
        this.userId = userId;
        this.turns = new ArrayList<>(turns);
        this.state = new HashMap<>(state);
        this.persistedTurnCount = persistedTurnCount;
    }

    public static AgentSession create(String userId) {
        return create(userId, null);
    }

    public static AgentSession create(String userId, String sessionId) {
        String id = sessionId == null || sessionId.isBlank()
                ? "sess_" + UUID.randomUUID().toString().replace("-", "")
                : sessionId;
        return new AgentSession(id, userId);
    }

    public void append(AgentTurn turn) {
        turns.add(turn);
    }

    public String sessionId() {
        return sessionId;
    }

    public String userId() {
        return userId;
    }

    public List<AgentTurn> turns() {
        return turns;
    }

    public Map<String, Object> state() {
        return state;
    }

    public int persistedTurnCount() {
        return persistedTurnCount;
    }

    public void markPersisted() {
        this.persistedTurnCount = turns.size();
    }

    public OptionalLong numericSessionId() {
        return parseLong(sessionId);
    }

    public OptionalLong numericUserId() {
        return parseLong(userId);
    }

    private static OptionalLong parseLong(String value) {
        try {
            return OptionalLong.of(Long.parseLong(value));
        } catch (NumberFormatException ignored) {
            return OptionalLong.empty();
        }
    }
}

File: src/session/agent-session.ts
import { userTurn, type AgentTurn } from "./agent-turn.js";
import { randomUUID } from "node:crypto";

export type SessionState = Record<string, unknown>;

export class AgentSession {
  readonly sessionId: string;
  readonly userId: string;

  private readonly turnList: AgentTurn[];
  private readonly sessionState: SessionState;
  private persistedTurnCountValue: number;

  constructor(input: {
    sessionId: string;
    userId: string;
    turns?: readonly AgentTurn[];
    state?: SessionState;
    persistedTurnCount?: number;
  }) {
    this.sessionId = input.sessionId;
    this.userId = input.userId;
    this.turnList = [...(input.turns ?? [])];
    this.sessionState = { ...(input.state ?? {}) };
    this.persistedTurnCountValue = input.persistedTurnCount ?? 0;
  }

  static create(userId: string, sessionId = createSessionId()): AgentSession {
    return new AgentSession({ sessionId, userId });
  }

  append(turn: AgentTurn): void {
    this.turnList.push(turn);
  }

  appendUser(content: string): void {
    this.append(userTurn(content));
  }

  turns(): readonly AgentTurn[] {
    return this.turnList;
  }

  state(): SessionState {
    return this.sessionState;
  }

  persistedTurnCount(): number {
    return this.persistedTurnCountValue;
  }

  markPersisted(): void {
    this.persistedTurnCountValue = this.turnList.length;
  }

  numericSessionId(): number | undefined {
    return parseSafeInteger(this.sessionId);
  }

  numericUserId(): number | undefined {
    return parseSafeInteger(this.userId);
  }
}

function createSessionId(): string {
  return `sess_${randomUUID().replaceAll("-", "")}`;
}

function parseSafeInteger(value: string): number | undefined {
  if (!/^-?\d+$/.test(value)) {
    return undefined;
  }

  const parsed = Number(value);
  return Number.isSafeInteger(parsed) ? parsed : undefined;
}
